File: sdks/typescript/examples/portfolio.ts
/**
 * O2 Exchange SDK - Portfolio Monitoring Example
 *
 * Balance and order monitoring:
 * - Streams balances and orders via WebSocket
 * - Displays formatted portfolio state
 * - Shows trade history
 *
 * Run: npx tsx examples/portfolio.ts
 */

import { type BalanceResponse, type Market, Network, O2Client } from "../src/index.js";

// ── Main ──────────────────────────────────────────────────────────

async function main() {
  const client = new O2Client({ network: Network.TESTNET });

  // Setup
  const wallet = client.generateWallet();
  console.log(`Wallet: ${wallet.b256Address}`);

  const { tradeAccountId } = await client.setupAccount(wallet);
  console.log(`Trade account: ${tradeAccountId}`);

  await sleep(3000);

  // Fetch markets for symbol mapping
  const markets = await client.getMarkets();
  console.log(`Markets: ${markets.map((m) => `${m.base.symbol}/${m.quote.symbol}`).join(", ")}`);

  // Display initial balances
  console.log("\n=== Portfolio Balances ===");
  const balances = await client.getBalances(tradeAccountId);
  for (const [symbol, balance] of Object.entries(balances)) {
    displayBalance(symbol, balance, markets);
  }

  // Display open orders
  console.log("\n=== Open Orders ===");
  for (const market of markets) {
    try {
      const orders = await client.getOrders(tradeAccountId, market, true, 50);
      if (orders.length > 0) {
        console.log(`\n${market.base.symbol}/${market.quote.symbol}:`);
        for (const order of orders) {
          const price = Number(order.price) / 10 ** market.quote.decimals;
          const qty = Number(order.quantity) / 10 ** market.base.decimals;
          const filled = order.quantity_fill
            ? Number(order.quantity_fill) / 10 ** market.base.decimals
            : 0;
          console.log(
            `  ${order.side} ${qty.toFixed(3)} @ ${price.toFixed(6)} ` +
              `(filled: ${filled.toFixed(3)}) [${order.order_id.slice(0, 10)}...]`,
          );
        }
      }
    } catch {
      // Skip markets with no orders
    }
  }

  // Stream balance updates
  console.log("\n=== Streaming Balance Updates ===");
  const balanceStream = await client.streamBalances(tradeAccountId);

  for await (const update of balanceStream) {
    console.log(`\n--- Balance Update (${new Date().toISOString()}) ---`);
    for (const entry of update.balance) {
      const symbol = findSymbol(entry.asset_id, markets) ?? entry.asset_id.slice(0, 10);
      const tradingBalance = Number(entry.trading_account_balance) / 10 ** 9;
      const locked = Number(entry.total_locked) / 10 ** 9;
      const unlocked = Number(entry.total_unlocked) / 10 ** 9;

      console.log(
        `  ${symbol}: available=${tradingBalance.toFixed(3)}, ` +
          `locked=${locked.toFixed(3)}, unlocked=${unlocked.toFixed(3)}`,
      );
    }
  }
}

// ── Helpers ───────────────────────────────────────────────────────

function displayBalance(symbol: string, balance: BalanceResponse, markets: Market[]) {
  const decimals = findDecimals(symbol, markets) ?? 9;
  const trading = Number(balance.trading_account_balance) / 10 ** decimals;
  const locked = Number(balance.total_locked) / 10 ** decimals;
  const unlocked = Number(balance.total_unlocked) / 10 ** decimals;
  const total = trading + locked + unlocked;

  if (total > 0) {
    console.log(
      `  ${symbol.padEnd(8)} | ` +
        `Available: ${trading.toFixed(3).padStart(12)} | ` +
        `Locked: ${locked.toFixed(3).padStart(12)} | ` +
        `Unlocked: ${unlocked.toFixed(3).padStart(12)} | ` +
        `Total: ${total.toFixed(3).padStart(12)}`,
    );
  }
}

function findDecimals(symbol: string, markets: Market[]): number | undefined {
  for (const m of markets) {
    if (m.base.symbol === symbol) return m.base.decimals;
    if (m.quote.symbol === symbol) return m.quote.decimals;
  }
  return undefined;
}

function findSymbol(assetId: string, markets: Market[]): string | undefined {
  for (const m of markets) {
    if (m.base.asset === assetId) return m.base.symbol;
    if (m.quote.asset === assetId) return m.quote.symbol;
  }
  return undefined;
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

main().catch(console.error);

File: sdks/rust/examples/portfolio.rs
use o2_sdk::crypto::*;
/// Portfolio monitoring example: streams balances and orders via WebSocket,
/// displays formatted portfolio state and P&L from trade history.
use o2_sdk::*;
use tokio_stream::StreamExt;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let mut client = O2Client::new(Network::Testnet);

    // Setup
    let wallet = client.generate_wallet()?;
    let owner_hex = to_hex_string(&wallet.b256_address);
    println!("Owner: {owner_hex}");

    let account = client.setup_account(&wallet).await?;
    let trade_account_id = account.trade_account_id.unwrap();
    println!("Trade account: {trade_account_id}");

    // Fetch markets and display initial balances
    let markets = client.get_markets().await?;
    println!("\n--- Available Markets ---");
    for m in &markets {
        println!(
            "  {} | min_order: {} | base_decimals: {} | quote_decimals: {}",
            m.symbol_pair(),
            m.min_order,
            m.base.decimals,
            m.quote.decimals
        );
    }

    // Display balances
    println!("\n--- Balances ---");
    let balances = client.get_balances(&trade_account_id).await?;
    for (symbol, bal) in &balances {
        let available = bal.trading_account_balance;
        let locked = bal.total_locked;
        let unlocked = bal.total_unlocked;
        println!("  {symbol}: available={available}, locked={locked}, unlocked={unlocked}");
    }

    // Display open orders for each market
    for market in &markets {
        let pair = market.symbol_pair();
        let orders_resp = client
            .get_orders(&trade_account_id, &pair, Some(true), 20)
            .await?;

        if !orders_resp.orders.is_empty() {
            println!("\n--- Open Orders ({pair}) ---");
            for order in &orders_resp.orders {
                let side = order.side;
                let price = order.price;
                let qty = order.quantity;
                let fill = order.quantity_fill.unwrap_or(0);
                let oid = &order.order_id;
                println!("  {side} {qty} @ {price} (filled: {fill}) [{oid}]");
            }
        }
    }

    // Display recent trades
    if let Some(market) = markets.first() {
        let pair = market.symbol_pair();
        let trades = client.get_trades(&pair, 10).await?;
        if !trades.trades.is_empty() {
            println!("\n--- Recent Trades ({pair}) ---");
            for trade in &trades.trades {
                let side = trade.side;
                let price = trade.price;
                let qty = trade.quantity;
                let ts = trade.timestamp;
                println!("  {side} {qty} @ {price} at {ts}");
            }
        }
    }

    // Stream balances in real-time
    println!("\n--- Streaming Balance Updates ---");
    let identity = Identity::ContractId(trade_account_id.to_string());
    let mut balance_stream = client.stream_balances(&[identity]).await?;

    while let Some(Ok(update)) = balance_stream.next().await {
        for entry in &update.balance {
            let asset = entry.asset_id.as_str();
            let available = entry.trading_account_balance;
            let locked = entry.total_locked;
            let unlocked = entry.total_unlocked;
            println!(
                "[balance] asset={} available={} locked={} unlocked={}",
                &asset[..10.min(asset.len())],
                available,
                locked,
                unlocked
            );
        }
    }

    Ok(())
}
